File: xwiki-platform-core/xwiki-platform-captcha/xwiki-platform-captcha-api/src/main/java/org/xwiki/captcha/Captcha.java
/*
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */
package org.xwiki.captcha;

import java.util.Map;

import org.xwiki.component.annotation.Role;
import org.xwiki.stability.Unstable;

/**
 * Describes a CAPTCHA that is able to display a challenge and validate and answer for the current context (i.e. user,
 * request, XWiki Context, etc.).
 *
 * @version $Id$
 * @since 10.8RC1
 */
@Unstable
@Role
public interface Captcha
{
    /**
     * @return the generated and displayed CAPTCHA challenge for the current context
     * @throws CaptchaException in case of problems
     */
    String display() throws CaptchaException;

    /**
     * @param parameters the configuration values to override when displaying the CAPTCHA. Missing parameters value are
     *            taken from the CAPTCHA's configuration
     * @return the generated and displayed CAPTCHA challenge for the current context using the given parameters to
     *         override the configuration
     * @throws CaptchaException in case of problems
     */
    String display(Map<String, Object> parameters) throws CaptchaException;

    /**
     * @return {@code true} if the CAPTCHA answer passed to the current context is valid; {@code false} otherwise
     * @throws CaptchaException in case of problems
     */
    boolean isValid() throws CaptchaException;

    /**
     * @param parameters the configuration values to override when validating the CAPTCHA's answer. Missing parameters
     *            value are taken from the CAPTCHA's configuration
     * @return {@code true} if the CAPTCHA answer passed to the current context, using the given parameters to override
     *         the configuration, is valid; {@code false} otherwise
     * @throws CaptchaException in case of problems
     */
    boolean isValid(Map<String, Object> parameters) throws CaptchaException;
}

File: xwiki-platform-core/xwiki-platform-captcha/xwiki-platform-captcha-jcaptcha/xwiki-platform-captcha-jcaptcha-webjar/src/main/webjar/captcha.js
/*
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */
require(['jquery', 'xwiki-events-bridge'], function($) {
  var registerListeners = function(container) {
    // If no container is specified, using the entire document as default.
    if (!container) {
      container = document;
    }

    // Click listener for the audio CAPTCHA play/listen button.
    $(container).find('.captcha-play').on('click', function() {
      $(this).siblings('audio.captcha-challenge')[0].play();
    });

    // Click listener for the CAPTCHA refresh/reload button.
    $(container).find('.captcha-refresh').on('click', function() {
      // We simply reset the 'src' attribute's value to make the CAPTCHA display a new challenge.
      var currentSrc = $(this).siblings('.captcha-challenge').prop('src');
      // Make sure to append a cache-buster to the src, since some elements might not reload if the src is the same
      // (e.g. image does not reload, but iframe reloads).
      $(this).siblings('.captcha-challenge')
        .prop('src', currentSrc.replace(/cache-buster=[0-9]+/gm, '') + 'cache-buster=' + new Date().getTime());
    });
  };

  // On load, register listeners for CAPTCHAs we can find.
  registerListeners(document);
  // Whenever a CAPTCHA is reloaded via AJAX, we need to re-register the listeners.
  $(document).on('xwiki:captcha:reloaded', function(event, data) {
    registerListeners(event.target);
  });

  // Does nothing but break Firefox's bfcache (back-forward cache) to make sure the CAPTCHA is reloaded when pressing
  // back and not showing a stale CAPTCHA.
  $(window).on('unload', function() {});
});